ddtrace/contrib/botocore/__init__.py

```diff
@@ -37,6 +37,21 @@
     Default: ``False``
 
 
+.. py:data:: ddtrace.config.botocore['s3_head_object_404_as_error']
+
+    Whether any ``botocore.exceptions.ClientError`` exceptions raise for 404 responses to
+    S3 ``HeadObject`` API calls should mark the representing span as an error.
+
+    By default we mark these spans as having an exception.
```

Comment:
Do we expect this to be the better default behavior and if so should we deprecate the current functionality of this integration? Any other cases of 404 that we need to add such a config option for?

Reply:
I am not sure.

I expect most people would prefer these not to be marked as errors by default, but I don't want to introduce that breaking change now.

> Any other cases of 404 that we need to add such a config option for?

I am not sure.

Another approach could be to allow users to provide an operation + status code list to ignore exceptions for?

```python
config.botocore.ignore_http_status_errors = {
    "s3.headobject": [404],
}
```

Reply:
I know it's not pretty but could this be handled using the current http config for `error_statuses`?

```
ddtrace.config.botocore.http_server.error_statuses = "500-599"
```

The botocore patch code would need to change to support this existing API but it could initially be treated as an unsupported/beta feature for this integration. While the `http_server` attribute might not be named well for this, such libraries that make calls to external services would need something similar to what we are doing here for s3.

Maybe that doesn't make much a difference in terms of our versioning and support policy compared to what you have here with a new configuration option. 

I would prefer a more general solution here than a one off for this specific use case. I am reluctant here to add a configuration option if it isn't what we think we'll need long term, even if we think the configuration would need some tweaking.

Reply:
yeah, I hadn't considered the more generic approach because I hadn't seen it from others before.

The existing code won't work, because we might need to set different policies per-operation, e.g. a 404 might be an error for `s3.getobject`, but not for `s3.headobject`

Reply:
@majorgreys I updated this PR with a more generic approach. I only updated the code in `patch.py` for now to get feedback before continuing.

This approach re-uses the existing `Config._HTTPServerConfig`. Maybe not my favorite approach, but easy enough to support the same behavior/range support.

what do you think of this approach?

Reply:
I think the more generic configuration is the right change. I am less sure about reusing the `Config._HTTPServerConfig` but I think it's fine for now since that's an internal API which does what we want here.

My earlier thought was reuse how we configure error statuses. That configuration is done globally. But given that in this integration we want to support the feature specific to the botocore operation, we'd need to add something new. 

What about `ddtrace.config.botocore.error_statuses` as `Dict[str, Config._HTTPServerConfig]` as you have it but not as an ignore list but as an include list as is the case with the global config?